datafusion/core/src/execution/context/mod.rs

```diff
@@ -1947,6 +1982,36 @@ impl SessionState {
             .await
     }
 
+    /// Create a [`PhysicalExpr`] from an [`Expr`] after applying type
+    /// coercion, and function rewrites.
+    ///
+    /// Note: The expression is not [simplified]
+    ///
+    /// # See Also:
+    /// * [`SessionContext::create_physical_expr`] for a higher-level API
```

Comment:
This is somewaht unrelated to this PR but It's news to me that "SessionContext" is higher level than "SessionState".  The only comparison between the two that I can find is https://docs.rs/datafusion/latest/datafusion/execution/context/struct.SessionContext.html#sessioncontext-sessionstate-and-taskcontext which says:

> A [SessionContext](https://docs.rs/datafusion/latest/datafusion/execution/context/struct.SessionContext.html) can be created from a [SessionConfig](https://docs.rs/datafusion/latest/datafusion/prelude/struct.SessionConfig.html) and stores the state for a particular query session. A single [SessionContext](https://docs.rs/datafusion/latest/datafusion/execution/context/struct.SessionContext.html) can run multiple queries.

> [SessionState](https://docs.rs/datafusion/latest/datafusion/execution/context/struct.SessionState.html) contains information available during query planning (creating [LogicalPlan](https://docs.rs/datafusion/latest/datafusion/logical_expr/enum.LogicalPlan.html)s and [ExecutionPlan](https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html)s).

From this it is not obvious that a "SessionContext" contains a "SessionState".  It's also very confusing when I would use "SessionContext" and when I would use "SessionState" in my user application.  Currently, I've been going off the philosophy of "if an API method needs a SessionState I'd better create one and if an API method needs a SessionContext I'd better create one."

Some kind of long form design documentation on these two types would be interesting (or if something already exists that I've missed or is in a blog post somewhere that would be cool too).

Reply:
I agree this is confusing -- I will write up some documentation about this

Reply:
Made https://github.com/apache/datafusion/pull/10350 to hopefully clarify